File: jclasslib/src/org/gjt/jclasslib/browser/detail/ConstantPoolDetailPane.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail;

import org.gjt.jclasslib.browser.*;
import org.gjt.jclasslib.browser.detail.constants.*;
import org.gjt.jclasslib.structures.*;
import org.gjt.jclasslib.structures.constants.*;

import javax.swing.*;
import javax.swing.tree.*;
import java.awt.*;
import java.util.*;

/**
    Detail pane showing constant pool entries. This class is a container for
    the classes defined in the <tt>constants</tt> subpackage and switches between
    the contained panes as required.
 
    @author <a href="mailto:jclasslib@gmx.net">Ingo Kegel</a>
    @version $Revision: 1.1.1.1 $ $Date: 2001-05-14 16:49:23 $
*/
public class ConstantPoolDetailPane extends AbstractDetailPane {

    private static final String SCREEN_CONSTANT_UTF8_INFO = "ConstantUtf8Info";
    private static final String SCREEN_CONSTANT_UNKNOWN = "ConstantUnknown";
    private static final String SCREEN_CONSTANT_CLASS_INFO = "ConstantClassInfo";
    private static final String SCREEN_CONSTANT_DOUBLE_INFO = "ConstantDoubleInfo";
    private static final String SCREEN_CONSTANT_LONG_INFO = "ConstantLongInfo";
    private static final String SCREEN_CONSTANT_FLOAT_INFO = "ConstantFloatInfo";
    private static final String SCREEN_CONSTANT_INTEGER_INFO = "ConstantIntegerInfo";
    private static final String SCREEN_CONSTANT_NAME_AND_TYPE_INFO = "ConstantNameAndTypeInfo";
    private static final String SCREEN_CONSTANT_STRING_INFO = "ConstantStringInfo";
    private static final String SCREEN_CONSTANT_REFERENCE = "ConstantReference";
    
    private HashMap constantTypeToDetailPane;
    
    public ConstantPoolDetailPane(BrowserInternalFrame parentFrame) {
        super(parentFrame);
    }

    protected void setupComponent() {
        setLayout(new CardLayout());
        constantTypeToDetailPane = new HashMap();
        JPanel pane;
        
        pane = new JPanel();
        pane.setBackground(Color.blue);
        add(pane, SCREEN_CONSTANT_UNKNOWN);
        
        addScreen(new ConstantUtf8InfoDetailPane(parentFrame),
            SCREEN_CONSTANT_UTF8_INFO);

        addScreen(new ConstantClassInfoDetailPane(parentFrame),
                  SCREEN_CONSTANT_CLASS_INFO);

        addScreen(new ConstantDoubleInfoDetailPane(parentFrame),
                  SCREEN_CONSTANT_DOUBLE_INFO);

        addScreen(new ConstantLongInfoDetailPane(parentFrame),
            SCREEN_CONSTANT_LONG_INFO);

        addScreen(new ConstantFloatInfoDetailPane(parentFrame),
            SCREEN_CONSTANT_FLOAT_INFO);

        addScreen(new ConstantIntegerInfoDetailPane(parentFrame),
            SCREEN_CONSTANT_INTEGER_INFO);

        addScreen(new ConstantNameAndTypeInfoDetailPane(parentFrame),
            SCREEN_CONSTANT_NAME_AND_TYPE_INFO);

        addScreen(new ConstantStringInfoDetailPane(parentFrame),
            SCREEN_CONSTANT_STRING_INFO);
                
        addScreen(new ConstantReferenceDetailPane(parentFrame),
            SCREEN_CONSTANT_REFERENCE);
                
    }
    
    public void show(TreePath treePath) {

        int constantPoolIndex = ((BrowserMutableTreeNode)treePath.getLastPathComponent()).getIndex();
        CPInfo constantPoolEntry = parentFrame.getClassFile().getConstantPool()[constantPoolIndex];
        
        String paneName = null;
        if (constantPoolEntry instanceof ConstantUtf8Info) {
            paneName = SCREEN_CONSTANT_UTF8_INFO;
        } else if (constantPoolEntry instanceof ConstantClassInfo) {
            paneName = SCREEN_CONSTANT_CLASS_INFO;
        } else if (constantPoolEntry instanceof ConstantDoubleInfo) {
            paneName = SCREEN_CONSTANT_DOUBLE_INFO;
        } else if (constantPoolEntry instanceof ConstantLongInfo) {
            paneName = SCREEN_CONSTANT_LONG_INFO;
        } else if (constantPoolEntry instanceof ConstantFloatInfo) {
            paneName = SCREEN_CONSTANT_FLOAT_INFO;
        } else if (constantPoolEntry instanceof ConstantIntegerInfo) {
            paneName = SCREEN_CONSTANT_INTEGER_INFO;
        } else if (constantPoolEntry instanceof ConstantNameAndTypeInfo) {
            paneName = SCREEN_CONSTANT_NAME_AND_TYPE_INFO;
        } else if (constantPoolEntry instanceof ConstantStringInfo) {
            paneName = SCREEN_CONSTANT_STRING_INFO;
        } else if (constantPoolEntry instanceof ConstantReference) {
            paneName = SCREEN_CONSTANT_REFERENCE;
        }

    
        CardLayout layout = (CardLayout)getLayout();
        if (paneName == null) {
            layout.show(this, SCREEN_CONSTANT_UNKNOWN);
        } else {
            AbstractDetailPane pane = (AbstractDetailPane)constantTypeToDetailPane.get(paneName);
            pane.show(treePath);
            layout.show(this, paneName);
        }
        
    }
    
    private void addScreen(JPanel pane, String name) {
        add(pane, name);
        constantTypeToDetailPane.put(name, pane);
    }
    
}

File: modules/browser/src/main/java/org/gjt/jclasslib/browser/detail/ConstantPoolDetailPane.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail

import org.gjt.jclasslib.browser.AbstractDetailPane
import org.gjt.jclasslib.browser.BrowserServices
import org.gjt.jclasslib.browser.BrowserTreeNode
import org.gjt.jclasslib.browser.detail.constants.*
import org.gjt.jclasslib.structures.Constant
import org.gjt.jclasslib.structures.ConstantType
import java.awt.CardLayout
import java.awt.Color
import java.util.*
import javax.swing.JPanel
import javax.swing.tree.TreePath

class ConstantPoolDetailPane(services: BrowserServices) : AbstractDetailPane(services) {

    private val constantTypeToDetailPane: EnumMap<ConstantType, AbstractDetailPane> = EnumMap(ConstantType::class.java)

    override fun setupComponent() {
        layout = CardLayout()

        add(JPanel().apply { background = Color.RED }, NAME_UNKNOWN)
        addCard(ConstantUtf8InfoDetailPane(browserServices), ConstantType.UTF8)
        addCard(ConstantClassInfoDetailPane(browserServices), ConstantType.CLASS)
        addCard(ConstantDoubleInfoDetailPane(browserServices), ConstantType.DOUBLE)
        addCard(ConstantLongInfoDetailPane(browserServices), ConstantType.LONG)
        addCard(ConstantFloatInfoDetailPane(browserServices), ConstantType.FLOAT)
        addCard(ConstantIntegerInfoDetailPane(browserServices), ConstantType.INTEGER)
        addCard(ConstantNameAndTypeInfoDetailPane(browserServices), ConstantType.NAME_AND_TYPE)
        addCard(ConstantStringInfoDetailPane(browserServices), ConstantType.STRING)
        addCard(ConstantReferenceDetailPane(browserServices), ConstantType.FIELDREF)
        addCard(ConstantReferenceDetailPane(browserServices), ConstantType.METHODREF)
        addCard(ConstantReferenceDetailPane(browserServices), ConstantType.INTERFACE_METHODREF)
        addCard(ConstantInvokeDynamicInfoDetailPane(browserServices), ConstantType.INVOKE_DYNAMIC)
        addCard(ConstantMethodHandleInfoDetailPane(browserServices), ConstantType.METHOD_HANDLE)
        addCard(ConstantMethodTypeDetailPane(browserServices), ConstantType.METHOD_TYPE)
    }

    override fun show(treePath: TreePath) {
        val constantPoolEntry = (treePath.lastPathComponent as BrowserTreeNode).element as Constant?
        if (constantPoolEntry != null) {
            val constantType = constantPoolEntry.constantType
            val detailPane = constantTypeToDetailPane[constantType]
            if (detailPane != null) {
                detailPane.show(treePath)
                showCard(constantType.name)
                return
            } else {
                showCard(NAME_UNKNOWN)

            }
        } else {
            showCard(NAME_UNKNOWN)
        }
    }

    private fun addCard(detailPane: AbstractDetailPane, constantType: ConstantType) {
        add(detailPane.displayComponent, constantType.name)
        constantTypeToDetailPane.put(constantType, detailPane)
    }

    private fun showCard(name: String) {
        (layout as CardLayout).show(this, name)
    }

    companion object {
        private val NAME_UNKNOWN = "ConstantUnknown"
    }

}

